Buildsystem and workflow improvements to the new Android custom builds

akien-mga opened the pull request:
- SCons: Generate `android_source.zip` during build

This is now needed after #27781, as this `android_source.zip` template
is used for custom Android builds from the editor.

I have yet to confirm that it works as expected though, my first test gave me a different and non-working APK when using a custom build with no custom config, compared to normal builds.

- Android: Improve dialogs about custom build template

Still not perfect but it makes things a bit clearer regarding the fact that it's going to install source code in the project folder for custom Android builds, as opposed to downloading the regular templates.

Fixes #28736 - at least to some extent, the rest of the confusion is simply due to no official templates existing yet with this `android_source.zip` file.

- Android: Extract libs from pre-built APKs when installing build template

First commit ignored the `libs` folder in the `android_source.zip`, but I learned that it's actually necessary. Instead of zipping it, we extract the libs directly from the pre-built APKs, so the `android_source.zip` stays 250 kiB instead of 54 MiB :)

---

Other things I plan to work on:

- Bring back the move of the generated APK to `bin/` when building the default templates. Need to find a way to have this behavior when building from Godot's repo, but not with custom builds. Any clue @godotengine/android @m4gr3d?
- Adjust docs for 3.2 (godotengine/godot-docs#2599).

akien-mga committed:
SCons: Generate android_source.zip during build

This is now needed after #27781, as this android_source.zip template
is used for custom Android builds from the editor.

akien-mga committed:
Android: Improve dialogs about custom build template

The language didn't make it clear that it's installing a *source* template
to the project folder, for later use when compiling custom APKs.

Fixes #28736.

akien-mga commented:
> I have yet to confirm that it works as expected though, my first test gave me a different and non-working APK when using a custom build with no custom config, compared to normal builds.

So according to @reduz the `libs` folder with the compiled libs for all arches and targets also need to be zipped... Which would make it impossible to do this via SCons, as each scons call builds one lib, so the zipping should happen at the end. Could be done via Gradle, but then it's kind of the same issue as the currently disabled moving of the generated APKs, we don't want it in custom builds.

I'll see if I can change the custom export system so that it extracts the libs directly from `android_release.apk` and `android_debug.apk`, instead of requiring them included in `android_source.zip`. That would save some disk space too for the templates.

neikeq commented:
Are you sure this would be impossible with SCons? Can't the zip be done with a command that runs after all lib commands are done? I don't know much about SCons, so can't tell how this is done, but I would be surprised if it wasn't possible.

akien-mga committed:
Android: Extract libs from pre-built APKs when installing build template

Otherwise we would need to include all of them in android_source.zip,
which means building the zip after all libs have been built by SCons
(so it would have to be done via gradle or a manual script).

By extracting it from the pre-built APKs, we save some disk space in
templates archives too.

akien-mga commented:
> Are you sure this would be impossible with SCons? Can't the zip be done with a command that runs after all lib commands are done? I don't know much about SCons, so can't tell how this is done, but I would be surprised if it wasn't possible.

Yes, because currently to built the 8 libs (4 arches, release + debug), we call SCons 8 times with different `android_arch` and `target` arguments. So only the 8th SCons call should actually create the `android_source.zip`, if it can find out somehow that all wanted libs have been built.

akien-mga commented:
> I'll see if I can change the custom export system so that it extracts the libs directly from `android_release.apk` and `android_debug.apk`, instead of requiring them included in `android_source.zip`. That would save some disk space too for the templates.

I've done that, it seems to work well.

akien-mga commented:
Amended the last commit to fix install path and permissions for the libs. Now I've tested the custom export successfully on my phone :)

akien-mga commented:
@m4gr3d That does sound better, but I'm not familiar enough with the Android build pipeline to implement that myself. This PR just adds some band aid on what @reduz implemented in #27781, and I assume he also doesn't know a lot more than that on how to build things for Android in a convenient way.

If you want to make a proposal for a better workflow using `aar` archives, that would definitely be welcome.

akien-mga commented on platform/android/SCsub:
```diff
@@ -55,3 +55,21 @@ if lib_arch_dir != '':
+# Change dir to avoid have zipped paths start from the android/java folder.
+olddir = os.getcwd()
+os.chdir(Dir('#platform/android/java').abspath)
+zipf = zipfile.ZipFile(File('#bin/android_source.zip').abspath, 'w', zipfile.ZIP_DEFLATED)
```
Looks like this runs early and fails if the `bin` folder wasn't created yet.
I'll wait to see if we go with another approach before reworking it.

m4gr3d commented:
> @m4gr3d That does sound better, but I'm not familiar enough with the Android build pipeline to implement that myself. This PR just adds some band aid on what @reduz implemented in #27781, and I assume he also doesn't know a lot more than that on how to build things for Android in a convenient way.
> 
> If you want to make a proposal for a better workflow using `aar` archives, that would definitely be welcome.

@akien-mga Sounds good. I'll prepare a PR using `aar` archives and send it within the next couple of days for review.

akien-mga commented:
@m4gr3d Awesome, thanks! I'll merge this PR nevertheless in the meantime, so that the slightly improved workflow is available in the upcoming 3.2-alpha1 in case we release it before your upcoming changes would be merged.
